File: solutions/other/indigo-code-2023/bracket.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
using namespace std;

bool balanced(string expr) {
    vector<char> stack;

    for (char c : expr) {
        if (c == '(') {
            stack.push_back(c);
        } else {
            if (stack.empty()) {
                return false;
            }
            char current_char = stack.back();
            stack.pop_back();
            if (current_char == '(' && c != ')') {
                return false;
            }
        }
    }

    return stack.empty();
}

int count_ = 0;

void recurse(string expr) {
    if (expr.find('?') == string::npos) {
        if (balanced(expr)) {
            count_++;
        }
        return;
    }

    int missing_idx = expr.find('?');
    if (std::count(expr.begin(), expr.begin() + missing_idx, ')') >
        std::count(expr.begin(), expr.begin() + missing_idx, '(')) {
        return;
    }
    
    string e1 = expr;
    string e2 = expr;

    e1[missing_idx] = ')';
    e2[missing_idx] = '(';

    recurse(e1);
    recurse(e2);
}

int main() {
    int n;
    cin >> n;
    string s;
    cin >> s;
    string e;
    for (char c : s) {
        e += c;
        e += '?';
    }
    recurse(e);
    cout << count_ << endl;
    return 0;
}

File: solutions/other/indigo-code-2023/bracket.py
n = int(input())
s = list(input().strip())
e = []
for c in s:
    e.append(c)
    e.append('?')
    
def balanced(expr):
    stack = []

    for char in expr:
        if char == "(":
            stack.append(char)
        else:
            if not stack:
                return False
            current_char = stack.pop()
            if current_char == '(':
                if char != ")":
                    return False
 
    if stack:
        return False
    return True

count = 0

def recurse(expr):
    global count
    if "?" not in expr:
        if balanced(expr):
            count += 1
        return
    
    missing_idx = expr.index('?')
    
    if expr[:missing_idx].count(')') > expr[:missing_idx].count('('):
        return
    
    e1 = list(expr)
    e2 = list(expr)
    
    e1[missing_idx] = ')'
    e2[missing_idx] = '('
    
    recurse(e1)
    recurse(e2)

recurse(e)
print(count)